### PR TITLE
feat: add gitlab provider templates

### PR DESCRIPTION
**Pull request by livio-a**
Adds possibility to manage and use GitLab (incl. self hosted) template based providers

```[tasklist]
### Definition of Ready
- [x] I am happy with the code
- [x] Short description of the feature/issue is added in the pr description
- [x] PR is linked to the corresponding user story
- [x] Acceptance criteria are met
- [x] All open todos and follow ups are defined in a new ticket and justified
- [x] Deviations from the acceptance criteria and design are agreed with the PO and documented.
- [x] No debug or dead code
- [x] Critical parts are tested automatically
- [ ] Where possible E2E tests are implemented
- [ ] Documentation/examples are up-to-date
- [x] All non-functional requirements are met
- [x] Functionality of the acceptance criteria is checked manually on the dev system.
```

relates to #4432

### CODE REVIEW

File: internal/idp/providers/oidc/oidc.go [hunk 1]
```diff
@@ -61,6 +62,13 @@ func WithRelyingPartyOption(option rp.Option) ProviderOpts {
 	}
 }
 
+// WithSelectAccount adds the select_account prompt to the auth request
+func WithSelectAccount() ProviderOpts {
```
Comment (adlerhurst) [inline internal/idp/providers/oidc/oidc.go:66]: is the option not provided by azure?

File: internal/domain/policy_login.go [hunk 2]
```diff
@@ -65,6 +67,35 @@ func (p IDPProvider) IsValid() bool {
...
+	case IDPTypeGitLab:
+		return "GitLab"
+	case IDPTypeGoogle:
+		return "Google"
+	case IDPTypeUnspecified,
+		IDPTypeOIDC,
+		IDPTypeJWT,
+		IDPTypeOAuth,
+		IDPTypeLDAP,
+		IDPTypeAzureAD,
```
Comment (adlerhurst) [inline internal/domain/policy_login.go:88]: isn't azure implemented?

File: internal/domain/idp.go [hunk 3]
```diff
@@ -36,12 +36,22 @@ const (
...
 		return "github"
+	case IDPTypeGitLab,
+		IDPTypeGitLabSelfHosted:
+		return "gitlab"
+	case IDPTypeUnspecified,
+		IDPTypeOIDC,
+		IDPTypeJWT,
+		IDPTypeOAuth,
+		IDPTypeLDAP,
+		IDPTypeAzureAD:
```
Comment (adlerhurst) [inline internal/domain/idp.go:53]: isn't azure implemented?

File: internal/query/projection/idp_template_test.go [hunk 4]
```diff
@@ -923,6 +923,512 @@ func TestIDPTemplateProjection_reducesGitHubEnterprise(t *testing.T) {
...
+			},
+			reduce: (&idpTemplateProjection{}).reduceGitLabIDPAdded,
+			want: wantReduce{
+				aggregateType:    eventstore.AggregateType("instance"),
+				sequence:         15,
+				previousSequence: 10,
+				executer: &testExecuter{
+					executions: []execution{
+						{
+							expectedStmt: "INSERT INTO projections.idp_templates3 (id, creation_date, change_date, sequence, resource_owner, instance_id, state, name, owner_type, type, is_creation_allowed, is_linking_allowed, is_auto_creation, is_auto_update) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14)",
```
Comment (adlerhurst) [inline internal/query/projection/idp_template_test.go:966]: shouldn't this statement be always the same for all provider types?

File: internal/query/idp_template.go [hunk 5]
```diff
@@ -350,6 +367,63 @@ var (
...
+	}
+	GitLabSelfHostedClientSecretCol = Column{
+		name:  projection.GitLabSelfHostedClientSecretCol,
+		table: gitlabSelfHostedIdpTemplateTable,
+	}
+	GitLabSelfHostedScopesCol = Column{
+		name:  projection.GitLabSelfHostedScopesCol,
+		table: gitlabSelfHostedIdpTemplateTable,
+	}
+)
```
Comment (adlerhurst) [inline internal/query/idp_template.go:426]:
```suggestion
)

```

File: internal/repository/idp/gitlab.go [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+		e.OptionChanges = options
+	}
+}
+
+func (e *GitLabIDPChangedEvent) Data() interface{} {
+	return e
+}
+
+func (e *GitLabIDPChangedEvent) UniqueConstraints() []*eventstore.EventUniqueConstraint {
+	// return []*eventstore.EventUniqueConstraint{NewAddGitLabIDPConfigNameUniqueConstraint(e.Name, e.Aggregate().ResourceOwner)} // TODO: ?
```
Comment (adlerhurst) [inline internal/repository/idp/gitlab.go:130]: todo

File: internal/repository/idp/gitlab.go [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+
+type GitLabSelfHostedIDPChangedEvent struct {
+	eventstore.BaseEvent `json:"-"`
+
+	ID           string              `json:"id"`
+	Name         *string             `json:"name,omitempty"`
+	Issuer       *string             `json:"issuer,omitempty"`
+	ClientID     *string             `json:"client_id,omitempty"`
+	ClientSecret *crypto.CryptoValue `json:"client_secret,omitempty"`
+	Scopes       []string            `json:"scopes,omitempty"` // TODO: tristate?
```
Comment (adlerhurst) [inline internal/repository/idp/gitlab.go:210]: todo

File: proto/zitadel/admin.proto [hunk 8]
```diff
@@ -1368,6 +1368,54 @@ service AdminService {
...
+        };
+
+        option (zitadel.v1.auth_option) = {
+            permission: "iam.idp.write"
+        };
+    }
+
+    // Add a new self hosted GitLab identity provider on the instance
+    rpc AddGitLabSelfHostedProvider(AddGitLabSelfHostedProviderRequest) returns (AddGitLabSelfHostedProviderResponse) {
+        option (google.api.http) = {
```
Comment (adlerhurst) [inline proto/zitadel/admin.proto:1397]: please add the openapi options

File: proto/zitadel/admin.proto [hunk 9]
```diff
@@ -4538,6 +4586,63 @@ message UpdateGitHubEnterpriseServerProviderResponse {
     zitadel.v1.ObjectDetails details = 1;
 }
 
+message AddGitLabProviderRequest {
+    // GitLab will be used as default, if no name is provided
+    string name = 1 [(validate.rules).string = {max_len: 200}];
```
Comment (adlerhurst) [inline proto/zitadel/admin.proto:4591]: @hifabienne please verify openapi options

File: proto/zitadel/idp.proto [hunk 10]
```diff
@@ -269,6 +269,8 @@ message ProviderConfig {
         JWTConfig jwt = 6;
         GitHubConfig github = 7;
         GitHubEnterpriseServerConfig github_es = 8;
+        GitLabConfig gitlab = 9;
```
Comment (adlerhurst) [inline proto/zitadel/idp.proto:272]: @hifabienne please verify openapi options

File: proto/zitadel/management.proto [hunk 11]
```diff
@@ -6584,6 +6584,54 @@ service ManagementService {
         };
     }
 
+    // Add a new GitLab identity provider in the organization
+    rpc AddGitLabProvider(AddGitLabProviderRequest) returns (AddGitLabProviderResponse) {
```
Comment (adlerhurst) [inline proto/zitadel/management.proto:6588]: please add openapi option

File: proto/zitadel/management.proto [hunk 12]
```diff
@@ -11212,6 +11260,63 @@ message UpdateGitHubEnterpriseServerProviderResponse {
     zitadel.v1.ObjectDetails details = 1;
 }
 
+message AddGitLabProviderRequest {
```
Comment (adlerhurst) [inline proto/zitadel/management.proto:11263]: please add openapi option

**Review by adlerhurst — CHANGES_REQUESTED** [review]
I'm asking about azure because there is a package called azure

**Comment by peintnermax** [comment]
@livio-a should the name attribute be required on self hosted or default to `Gitlab`?

File: internal/idp/providers/gitlab/gitlab_test.go [hunk 13]
```diff
@@ -33,7 +33,7 @@ func TestProvider_BeginAuth(t *testing.T) {
 				scopes:       []string{"openid"},
 			},
 			want: &oidc.Session{
-				AuthURL: "https://gitlab.com/oauth/authorize?client_id=clientID&prompt=select_account&redirect_uri=redirectURI&response_type=code&scope=openid&state=testState",
+				AuthURL: "https://gitlab.com/oauth/authorize?client_id=clientID&redirect_uri=redirectURI&response_type=code&scope=openid&state=testState",
```
[the remark this answers is not in the document]
Comment (livio-a) [reply internal/idp/providers/gitlab/gitlab_test.go:36]: implemented

File: internal/domain/policy_login.go [hunk 14]
```diff
@@ -65,6 +67,35 @@ func (p IDPProvider) IsValid() bool {
...
+		IDPTypeGitHubEnterprise,
+		IDPTypeGitLabSelfHosted:
+		fallthrough
+	default:
+		// we should never get here, so log it
+		logging.Errorf("name of provider (type %v) is empty: %v", p.IDPConfigID, p.Type)
```
[the remark this answers is not in the document]
Comment (livio-a) [reply internal/domain/policy_login.go]: fixed (with correct field)

File: internal/command/org_idp_model.go [hunk 15]
```diff
@@ -446,6 +446,145 @@ func (wm *OrgGitHubEnterpriseIDPWriteModel) NewChangedEvent(
...
+				ResourceOwner: orgID,
+			},
+			ID: id,
+		},
+	}
+}
+
+func (wm *OrgGitLabIDPWriteModel) Reduce() error {
+	return wm.GitLabIDPWriteModel.Reduce()
+}
```
[the remark this answers is not in the document]
Comment (livio-a) [reply internal/command/org_idp_model.go]: removed

File: internal/command/org_idp_model.go [hunk 16]
```diff
@@ -446,6 +446,145 @@ func (wm *OrgGitHubEnterpriseIDPWriteModel) NewChangedEvent(
...
+				ResourceOwner: orgID,
+			},
+			ID: id,
+		},
+	}
+}
+
+func (wm *OrgGitLabSelfHostedIDPWriteModel) Reduce() error {
+	return wm.GitLabSelfHostedIDPWriteModel.Reduce()
+}
```
[the remark this answers is not in the document]
Comment (livio-a) [reply internal/command/org_idp_model.go]: removed

**Review by adlerhurst — COMMENTED** [review]
only errors remaining

**Review by adlerhurst — APPROVED** [review]

File: proto/zitadel/admin.proto (hunk 8 above)
Comment (adlerhurst) [reply proto/zitadel/admin.proto:1397]: will be resolved in https://github.com/zitadel/zitadel/issues/5439

File: proto/zitadel/idp.proto (hunk 10 above)
Comment (adlerhurst) [reply proto/zitadel/idp.proto:272]: will be resolved in https://github.com/zitadel/zitadel/issues/5439

File: proto/zitadel/management.proto (hunk 11 above)
Comment (adlerhurst) [reply proto/zitadel/management.proto:6588]: will be resolved in https://github.com/zitadel/zitadel/issues/5439

File: proto/zitadel/management.proto (hunk 12 above)
Comment (adlerhurst) [reply proto/zitadel/management.proto:11263]: will be resolved in https://github.com/zitadel/zitadel/issues/5439

**Review by adlerhurst — APPROVED** [review]

**Pull request merged**